sycl/include/CL/sycl/group_algorithm.hpp

```diff
@@ -0,0 +1,839 @@
+    T>
+joint_reduce(Group g, Ptr first, Ptr last, T init, BinaryOperation binary_op) {
+  // FIXME: Do not special-case for half precision
+  static_assert(
+      std::is_same<decltype(binary_op(init, *first)), T>::value ||
+          (std::is_same<T, half>::value &&
+           std::is_same<decltype(binary_op(init, *first)), float>::value),
+      "Result type of binary_op must match reduction accumulation type.");
+#ifdef __SYCL_DEVICE_ONLY__
+  T partial = sycl::known_identity_v<BinaryOperation, T>;
+  sycl::detail::for_each(
+      g, first, last, [&](const typename detail::remove_pointer<Ptr>::type &x) {
+        partial = binary_op(partial, x);
+      });
+  return reduce_over_group(g, partial, init, binary_op);
+#else
+  (void)g;
+  (void)last;
+  throw runtime_error("Group algorithms are not supported on host device.",
+                      PI_INVALID_DEVICE);
+#endif
+}
+
+// ---- any_of_group
+template <typename Group>
+detail::enable_if_t<is_group_v<std::decay_t<Group>>, bool>
+any_of_group(Group, bool pred) {
+#ifdef __SYCL_DEVICE_ONLY__
+  return sycl::detail::spirv::GroupAny<Group>(pred);
+#else
+  (void)pred;
+  throw runtime_error("Group algorithms are not supported on host device.",
+                      PI_INVALID_DEVICE);
+#endif
+}
+
+template <typename Group, typename T, class Predicate>
+detail::enable_if_t<is_group_v<Group>, bool> any_of_group(Group g, T x,
+                                                          Predicate pred) {
+  return any_of_group(g, pred(x));
+}
+
+// ---- joint_any_of
+template <typename Group, typename Ptr, class Predicate>
+detail::enable_if_t<
+    (is_group_v<std::decay_t<Group>> && detail::is_pointer<Ptr>::value), bool>
+joint_any_of(Group g, Ptr first, Ptr last, Predicate pred) {
+#ifdef __SYCL_DEVICE_ONLY__
+  using T = typename detail::remove_pointer<Ptr>::type;
+  bool partial = false;
+  sycl::detail::for_each(g, first, last, [&](T &x) { partial |= pred(x); });
+  return any_of_group(g, partial);
+#else
+  (void)g;
+  (void)first;
+  (void)last;
+  (void)pred;
+  throw runtime_error("Group algorithms are not supported on host device.",
+                      PI_INVALID_DEVICE);
+#endif
+}
+
+// ---- all_of_group
+template <typename Group>
+detail::enable_if_t<is_group_v<std::decay_t<Group>>, bool>
+all_of_group(Group, bool pred) {
+#ifdef __SYCL_DEVICE_ONLY__
+  return sycl::detail::spirv::GroupAll<Group>(pred);
+#else
+  (void)pred;
+  throw runtime_error("Group algorithms are not supported on host device.",
+                      PI_INVALID_DEVICE);
+#endif
+}
+
+template <typename Group, typename T, class Predicate>
+detail::enable_if_t<is_group_v<std::decay_t<Group>>, bool>
+all_of_group(Group g, T x, Predicate pred) {
+  return all_of_group(g, pred(x));
+}
+
+// ---- joint_all_of
+template <typename Group, typename Ptr, class Predicate>
+detail::enable_if_t<
+    (is_group_v<std::decay_t<Group>> && detail::is_pointer<Ptr>::value), bool>
```

Comment:
```suggestion
    is_group_v<std::decay_t<Group>>, bool>
```

Reply:
@Pennycook , are you sure about this? You are certainly correct that the spec does not require `Ptr` to be `is_pointer`.  But all the other `joint_`  variants ( like `joint_reduce` ) do have that requirement.

For reference, imagine the user just uses an arbitrary non-pointer type for Ptr. 
```int x = 3, y=4;
bool hasEven = joint_any_of(it.get_group(), x, y, evenp);
```

If I attempt to compile this while we keep the `is_pointer<Ptr>` guard we have now, then I get 
```error: no matching function for call to 'joint_any_of'
 note: candidate template ignored: requirement 'detail::is_pointer<int>::value' was not satisfied [with Group = sycl::group<1>, Ptr = int, Predicate = (lambda at all_any_none.cpp:50:26)]
joint_any_of(Group g, Ptr first, Ptr last, Predicate pred) {
``` 
That's pretty clear (for C++).

But if we remove this guard as you suggest, then we get two compiler errors. They still force the user to use a pointer. 
```
error: indirection requires pointer operand ('int' invalid)
    f(*p);
      ^~
/iusers/cperkins/sycl_workspace/build/bin/../include/sycl/CL/sycl/group_algorithm.hpp:298:17: note: in instantiation of function template specialization 'sycl::detail::for_each<sycl::group<1>, int, (lambda at /iusers/cperkins/sycl_workspace/build/bin/../include/sycl/CL/sycl/group_algorithm.hpp:298:42)>' requested here
  sycl::detail::for_each(g, first, last, [&](T &x) { partial |= pred(x); });


error: SYCL kernel cannot call an undefined function without SYCL_EXTERNAL attribute
  sycl::detail::for_each(g, first, last, [&](T &x) { partial |= pred(x); });
                ^
/iusers/cperkins/sycl_workspace/build/bin/../include/sycl/CL/sycl/group_algorithm.hpp:99:10: note: 'for_each<sycl::group<1>, int, (lambda at /iusers/cperkins/sycl_workspace/build/bin/../include/sycl/CL/sycl/group_algorithm.hpp:298:42)>' declared here
Function for_each(Group g, Ptr first, Ptr last, Function f) {
```

I guess it's fine, but just surprising (to me).

Reply:
Hm, you're right.

I thought that the dereference would happen inside of `Predicate`, but that's not true.  What I said explains why `joint_*_of` functions don't constrain their arguments to pointers to _fundamental_ types (like the other `joint_*` functions do), but your argument that we should still constrain their arguments to pointers makes sense.

I'm happy for things to be merged as they are, with similar justification to before (i.e. we're currently more restrictive than the specification, not less).  I'll add this to the list of questions to discuss at Khronos.

FYI - I've opened https://gitlab.khronos.org/sycl/Specification/-/issues/554 if you want to follow along.